### PR TITLE
ISPN-6906 Add ExternalPojo marker

### PR DESCRIPTION
**Pull request by galderz**
* Marker interface used to signal test classes that are supposed to be
  externally marshalled.

### CODE REVIEW

**Comment by pruivo** [comment]
is this exception supposed to printed? 
```
Caused by: java.lang.AssertionError: Check support for: class org.infinispan.tx.exception.ExceptionInCommandTest$MyDelta
```

File: core/src/main/java/org/infinispan/marshall/core/ExternallyMarshallable.java [hunk 1]
```diff
@@ -2,18 +2,18 @@
...
    }
 
-   public static boolean isAllowed(Object obj) {
+   static boolean isAllowed(Object obj) {
       String pkg = obj.getClass().getPackage().getName();
       boolean isBlackList =
             obj instanceof Serializable
             && isMarshallablePackage(pkg)
+            && !(obj instanceof ExternalPojo)
             && !isWhiteList(obj.getClass().getName());
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/marshall/core/ExternallyMarshallable.java]: I'm not sure how this is supposed to work for user classes, as users can't modify the whitelist and they can't always modify their classes to extend `ExternalPojo` either. They could write+configure externalizers, but do we want to force them to write externalizers for all their classes?

My thinking was that we'd only have the blacklist active during the tests, so we'd only need `ExternalPojo` for our internal classes.

Comment (pruivo) [inline core/src/main/java/org/infinispan/marshall/core/ExternallyMarshallable.java]: hmmm... what is the problem that ```ExternalPojo``` tries to solve?

**Comment by pruivo** [comment]
the ```HotRodDistributedEventsTest``` failures seems related to missing ```ExternalPojo``` in some test classes.

**Comment by rvansa** [comment]
Why do you use a marker interface instead of annotation? I thought marker interfaces are Java 1.4 thing.

File: core/src/main/java/org/infinispan/marshall/core/ExternallyMarshallable.java (hunk 1 above)
Comment (galderz) [inline core/src/main/java/org/infinispan/marshall/core/ExternallyMarshallable.java]: > I'm not sure how this is supposed to work for user classes

^ That was never the objective of this, but only for test classes and the check only happens when assertions are enabled.

Comment (galderz) [inline core/src/main/java/org/infinispan/marshall/core/ExternallyMarshallable.java]: > hmmm... what is the problem that ExternalPojo tries to solve?

Detect Infinispan classes for which we should have externalizers. If we forget that a class has a externalizer, and it uses standard Serialization, this check will catch it.

If we really expect it to be Serializable, e.g. testing Serializable classes, then the class can be made to extend ExternalPojo interface as marker.

**Comment by galderz** [comment]
> the HotRodDistributedEventsTest failures seems related to missing ExternalPojo in some test classes.

Ok, I'll address that.

> Why do you use a marker interface instead of annotation? I thought marker interfaces are Java 1.4 thing.

Only used for testing, either method should work just fine.

**Comment by galderz** [comment]
Fixed `ExceptionInCommandTest` and `HotRodDistributedEventsTest`

**Comment by galderz** [comment]
Rebasing...

**Comment by galderz** [comment]
Rebased

File: core/src/main/java/org/infinispan/marshall/core/ExternallyMarshallable.java (hunk 1 above)
Comment (danberindei) [inline core/src/main/java/org/infinispan/marshall/core/ExternallyMarshallable.java]: @galderz I was thinking about the `ExternalPojo` javadoc, which says `One such example are pojos used in testing, but not limited to,` and I wasn't sure what the actual limits are ;)

**Comment by danberindei** [comment]
Integrated, thanks Galder!

**Pull request closed without merging**